User defined hashmap key equality function

lucaato opened the pull request:
Following issue #76238 I wanted to give an example of this possibly breaking change. It could be made into a nonbreaking change if needed.
I think one of the main reasons this didn't get implemented is that embedded systems usually don't deal with large enough amounts of data for a hashmap to make sense.
But if the feature is not too invasive and does not create problems this could be useful to some.
Thanks in advance for taking the time to review the PR.

cfriedt commented on include/zephyr/sys/hash_map.h:
```diff
@@ -102,6 +104,16 @@ extern "C" {
+/*
+ * Default equality function used by the hash map implementations
+ */
+static inline bool sys_hashmap_default_equal(uint64_t left, uint64_t right) {
```
Something is inconsistent between the associated issue and this PR.

Namely, the issue describes using strings as _keys_ for hash maps, but it would appear that this PR aims to use strings as _values_. I might be mistaken though.

Using strings as _values_ should be considerably easier to implement, as the pointer of course will fit inside the 64-bit value for both 32-bit and 64-bit architectures, and I believe that is supported today.

However, the value side of the hash map does not need to be unique. Only the key of the hash map needs to be unique.

If the aim of this PR is to use strings as keys, then I would say a significant problem is that hash maps currently only use 32-bit keys. I.e. on 64-bit architectures, the 64-bit `char *` would not be a valid key. So there would be a need to define 64-bit hash functions, maps, etc.

However, again, with the concept of using strings as keys simply requires not using duplicate keys. So perhaps it would be easier to de-duplicate strings first?

What is a simple way to de-duplicate strings? Well, calculating a 32-bit hash of the string and seeing if the key exists in some collection is one way. Another way is to use a trie. I think I would lean toward the former if the strings were expected to be mostly dissimilar. 

This is typically what would be done in languages that implicitly ref-count strings such as C++ (and others), have smart pointers, automatic destruction, etc.

What might be convenient to do in this case is to define a separate specialization of the hash map just for C strings, since there is that additional step of deduplicating keys (and actually, one way to define a set is to use a hash map).

Maybe a `string_set` and some method like add, remove, contains, and hash or something like that?

I would guess that a key space of 4B is large enough to handle all strings used in Zephyr for most practical applications.

lucaato replied:
What I'm trying to do is make it feasible to store generic data types in the key field of hashmaps of course through the use of pointers. This should still be achieved by the user who is required to provide a custom equality function and a custom hash function that correctly hashes the associated pointer.
I don't get what you're saying about the 32-bit hash not being enough, it's true that currently, the hash function returns a 32-bit number but it seems to me that the key is still stored as an [`uint64_t`](https://github.com/zephyrproject-rtos/zephyr/blob/5305aa615b0f800508e253c3398712eca1b209b4/lib/hash/hash_map_sc.c#L20) so it seems completely possible to store a pointer even if the hash function returns a smaller hash.
I see you also mentioned handling the allocations, and yes that makes things more complicated to handle on the hashmap code part. Currently however users are already storing pointers in the value field of maps and when they do they also handle deallocating them.
A valid use case I see is something like the following:
```c
#include <zephyr/sys/hash_function.h>
#include <zephyr/sys/hash_map.h>
#include <zephyr/sys/hash_map_api.h>
#include <zephyr/sys/hash_map_sc.h>

// NOLINTNEXTLINE(hicpp-signed-bitwise)
const struct sys_hashmap_config object_map_config
    = SYS_HASHMAP_CONFIG(OBJECT_MAP_MAX_ENTRIES, SYS_HASHMAP_DEFAULT_LOAD_FACTOR);

// we check that we can cast correctly to uint64_t
BUILD_ASSERT(sizeof(uintptr_t) <= sizeof(uint64_t), "A pointer is larger than 64 bits ?!");

// NOLINTNEXTLINE(hicpp-signed-bitwise)
const struct sys_hashmap_config object_map_config
    = SYS_HASHMAP_CONFIG(OBJECT_MAP_MAX_ENTRIES, SYS_HASHMAP_DEFAULT_LOAD_FACTOR);

void example(...) {
    struct sys_hashmap_data entries_map_data = {};
    struct sys_hashmap entries_map = {
        .api = &sys_hashmap_sc_api,
        .config = &object_map_config,
        .data = &entries_map_data,
        // only pointer to valid c strings should be hashed
        .hash_func = hash_string,
        .eq_func = eq_string,
        .alloc_func = SYS_HASHMAP_DEFAULT_ALLOCATOR,
    };

    // use the map

exit:
    sys_hashmap_clear(&entries_map, NULL, NULL);
    return result;
}

static uint32_t hash_string(const void *uint64_ptr_key, size_t len) {
    CHECK_HALT(len != 8, "Received an invalid length for the hasmap key, expected only uint64_t");

    uint64_t key_pointer = {};
    memcpy(&key_pointer, uint64_ptr_key, sizeof(uint64_t));

    // we know (and hope) this is actually a string
    // NOLINTNEXTLINE(performance-no-int-to-ptr)
    const char *key_string = UINT_TO_POINTER(key_pointer);
    size_t key_string_len = strlen(key_string);

    //LOG_INF("Hashing '%s' = '%016x'", key_string, sys_hash32(key_string, key_string_len));

    return sys_hash32(key_string, key_string_len);
}

static bool eq_string(uint64_t key_left, uint64_t key_right) {
    const char *key_left_string = UINT_TO_POINTER(key_left);
    const char *key_right_string = UINT_TO_POINTER(key_right);

    return strcmp(key_left_string, key_right_string) == 0;
}
```
The hashmap does not handle anything besides what it always did but by exposing the `eq_func` field the user can override the default comparison logic and compare the keys.
What I got about not using duplicated strings is that if all the pointers used to store the string `"example"` pointed to the same instance (same pointer) we wouldn't need any fancy equality function but this is more complicated in other ways.
Currently, I do not see why the simple equality function override method would not work, but I'm new to the field and I'm probably missing something.
For the same string, you get the same hash (given that the hash function is defined correctly) after that the correct bucket will be checked (in the case of sc) every entry key stored in the bucket will be compared to the key passed (another pointer that could point to the same string) using the equality function.

cfriedt replied:
> What I'm trying to do is make it feasible to store generic data types in the key field of hashmaps of course through the use of pointers. This should still be achieved by the user who is required to provide a custom equality function and a custom hash function that correctly hashes the associated pointer.
> I don't get what you're saying about the 32-bit hash not being enough, it's true that currently, the hash function returns a 32-bit number but it seems to me that the key is still stored as an [`uint64_t`](https://github.com/zephyrproject-rtos/zephyr/blob/5305aa615b0f800508e253c3398712eca1b209b4/lib/hash/hash_map_sc.c#L20) so it seems completely possible to store a pointer even if the hash function returns a smaller hash.

That's correct - sorry, I was looking at this originally.

https://github.com/zephyrproject-rtos/zephyr/blob/main/include%2Fzephyr%2Fsys%2Fhash_map.h#L132-L133

It looks like the key type was left as uint64_t intentionally. It's been a while.

> I see you also mentioned handling the allocations, and yes that makes things more complicated to handle on the hashmap code part. Currently however users are already storing pointers in the value field of maps and when they do they also handle deallocating them.

As long as key allocations are handled externally, that should be ok. Currently only value allocations are managed.

> The hashmap does not handle anything besides what it always did but by exposing the `eq_func` field the user can override the default comparison logic and compare the keys.

Sure - I think it should be doable. Can you make the necessary changes in this PR?

There would need to be new tests added.

lucaato replied:
Ok, thanks again for your time. I will try to clean the pr up and add a few tests, maybe even some additions to the documentation would be a good idea.

cfriedt reviewed (commented):
Should this be converted to a discussion? It seems like the pull request isn't intended to land, but just to illustrate some thoughts.

lucaato commented:
> Should this be converted to a discussion? It seems like the pull request isn't intended to land, but just to illustrate some thoughts.

Yes this is more of a demonstration of a possible solution than an actual pr this could be a discussion.

andyross reviewed (changes requested):
Missed this the first time around.  Can you split this work between a PR with the uncontroversial doc/naming/test/formatting fixups (but please isolate the latter! there's a lot of whitespace mixed up with semantic changes in this PR) and one adding the new eq_func feature?

lucaato committed:
lib: hash: Hash map module docstring fixups

Applied small fixes to the hash map module docstrings.

Signed-off-by: Luca Arato <luca.arato@secomind.com>

lucaato committed:
lib: hash: Added user defined hash map equality

Allows storing arbitrary types in the key of the
hash map by adding a user defined equality function

Signed-off-by: Luca Arato <luca.arato@secomind.com>

lucaato committed:
lib: hash: Added test specific to eq_func usage

Added a variation of the hash map tests to specifically
target the usage of the introduced eq_func

Signed-off-by: Luca Arato <luca.arato@secomind.com>

lucaato commented:
> Missed this the first time around. Can you split this work between a PR with the uncontroversial doc/naming/test/formatting fixups (but please isolate the latter! there's a lot of whitespace mixed up with semantic changes in this PR) and one adding the new eq_func feature?

I removed all the whitespace changes and kept all the docstring format changes in the first commit. The other two commits contain changes related to the eq_func feature. There's way less noise now and reviewing should be easier.

The docstring changes are still included in this pr but I can still split them if needed.

cfriedt commented:
@lucaato - please go through the failed tests in CI and correct any errors. All checks must pass / be green for the PR to be merged.

E.g. compliance failures

https://github.com/zephyrproject-rtos/zephyr/actions/runs/11068473415/job/30753988176#step:11:1

cfriedt commented:
@lucaato - very close!

There are still a few code formatting errors being raised.

https://github.com/zephyrproject-rtos/zephyr/actions/runs/11147858653/job/30983167761#step:11:1

IIRC, we have a git commit pre-hook that will catch those before they are checked-in.

E.g.
https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style
https://cgnd.dev/posts/enforce-zephyr-code-quality-pre-commit/

cfriedt commented on tests/lib/hash_map_string/src/_main.c:
```diff
@@ -0,0 +1,125 @@
+#include "zephyr/sys/util.h"
+
+static uint32_t hash_string(const void *uint64_ptr_key, size_t len);
+static bool eq_string(uint64_t key_left, uint64_t key_right);
```
I think we can probably get away with not duplicating the entire `hash_map` testsuite.

What about creating a separate testsuite inside the `tests/lib/hash_map` app with a descriptive comment or two?

E.g.

```cpp
ZTEST_SUITE(hash_map_custom_eq_func, NULL, setup, NULL, after, NULL);
```

lucaato replied:
That makes sense, do you prefer that I conditionally compile just lines of the functions or can I just conditionally compile the whole function in the same test suite ?
Thanks in advance.

cfriedt replied:
Probably easiest to conditionally compile one ZTEST in the same testsuite. YMMV though.

cfriedt replied:
Whatever the more minimal change is. It's hard for me to say at the moment.

lucaato replied:
Hi again @cfriedt, I tried applying the requested changes to the `_main.c` and `get.c` test files to give you an idea of what it would look like if I were to inline the `ifdef`s. It is probably a smaller change than cloning the whole function but it is also harder to read IMHO. Do you prefer it like this or cloning the whole function is ok to you ?

lucaato commented on tests/lib/hash_map_string/src/_main.c:
```diff
@@ -0,0 +1,125 @@
+	const char *key_left_string = UINT_TO_POINTER(key_left);
+	const char *key_right_string = UINT_TO_POINTER(key_right);
+
+	return strcmp(key_left_string, key_right_string) == 0;
```
The notion about specific management of string as keys surely make sense. After a way to define a custom equality function a string map provided by the zephyr API could be implemented I think.

lucaato replied:
I should have adhered to the supported `strcmp` usage.

cfriedt reviewed (changes requested):
I think it would be much easier to make the string map API public and implemented in a header file as static inline functions. It should be quite useful though and is mostly a wrapper around the existing hash map API.

W.r.t. the hash map testsuite, it would be better to define two separate datasets (or keysets, I guess?); one for integers and another for substrings of lorem ipsum. Both can be declared `extern` in `_main.h` and put into `_main.c`.

The integer dataset should be in bss and be initialized in `setup()` as numbers `[0, MANY-1]`.

The indexed allocator idea is good, because all that the string dataset needs to include really are begin and end offsets for strings from a single constant lorem ipsum instance. So just have an array of size `MANY` that includes pre-selected substring indices.

Another macro can be made to refer to either the integer dataset or the string dataset.

There may be some special scenarios where invalid values should be used, and again, there should be macros used to refer to either the integer or string values.

With that, most of the test suite can be left unmodified.

Overall, looks really good 👍 Thank you for all of the work you have put into this @lucaato !

cfriedt commented:
@lucaato - this could use a rebase. It's clue to done as well. Just needs to be pushed past the finish line 🏁

lucaato commented:
Hi again, thanks for your patience with this PR :smile: I applied some of the changes you suggested to the code, and now the test looks much better. I still need to do some cleanup, but it should almost be ready for another review. Thanks again.

lucaato commented on tests/lib/hash_map/src/get.c:
```diff
@@ -37,13 +37,15 @@ ZTEST(hash_map, test_get_false)
 	}
 
-	zassert_false(sys_hashmap_get(&map, 0x4242424242424242ULL, NULL));
+	KEY_TYPE key = ALLOC_KEY(0x4242424242424242ULL);
+	zassert_false(GET_FUNC_NAME(&map, key, NULL));
+	FREE_KEY(key);
```
I'm still not sure about these parts of the test, it does not look too good but it works.

lucaato commented:
This should be almost ready for another review, @cfriedt if you'd like to you can give it a go. Thanks in advance :smile:

cfriedt commented on tests/lib/hash_map/testcase.yaml:
```diff
@@ -13,11 +13,23 @@ tests:
+      - CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=8192
+      - CONFIG_SYS_HASH_MAP_CHOICE_SC=y
+      - CONFIG_SYS_HASH_FUNC32_CHOICE_DJB2=y
+      - CONFIG_TEST_USE_CUSTOM_EQ_FUNC=y
```
```suggestion
      - CONFIG_TEST_USE_CUSTOM_EQ_FUNC=y
      - CONFIG_BASE64=y
```

cfriedt commented on tests/lib/hash_map/testcase.yaml:
```diff
@@ -13,11 +13,23 @@ tests:
+      - CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE=8192
+      - CONFIG_SYS_HASH_MAP_CHOICE_OA_LP=y
+      - CONFIG_SYS_HASH_FUNC32_CHOICE_DJB2=y
+      - CONFIG_TEST_USE_CUSTOM_EQ_FUNC=y
```
```suggestion
      - CONFIG_TEST_USE_CUSTOM_EQ_FUNC=y
      - CONFIG_BASE64=y
```

cfriedt commented on tests/lib/hash_map/prj.conf:
```diff
@@ -7,4 +7,6 @@ CONFIG_ZTEST=y
 CONFIG_SYS_HASH_FUNC32=y
 CONFIG_SYS_HASH_MAP=y
 
+CONFIG_BASE64=y
+
```
```suggestion
```

cfriedt commented on tests/lib/hash_map/Kconfig:
```diff
@@ -21,4 +21,13 @@ config TEST_LIB_HASH_MAP_MAX_ENTRIES
 
+config TEST_USE_CUSTOM_EQ_FUNC
+	bool "Use a custom equality function"
+	default n
```
```suggestion
```

cfriedt reviewed (commented):
Some minor suggestions

cfriedt commented:
@lucaato - there are some compliance errors in terms of code formatting but it looks OK to me otherwise.
